diff -Nur mysql-5.5-5.5.15.orig/mysql-test/t/disabled.def mysql-5.5-5.5.15/mysql-test/t/disabled.def
--- mysql-5.5-5.5.15.orig/mysql-test/t/disabled.def	2011-07-13 21:09:03.000000000 +0200
+++ mysql-5.5-5.5.15/mysql-test/t/disabled.def	2011-08-08 11:37:31.634672359 +0200
@@ -16,3 +16,5 @@
 create-big               : Bug#11748731 2010-11-15 mattiasj was not tested
 archive-big              : Bug#11817185 2011-03-10 Anitha Disabled since this leads to timeout on Solaris Sparc
 log_tables-big           : Bug#11756699 2010-11-15 mattiasj report already exists
+partition_rename_longfilename : Fails when building with sbuild and schroots
+file_contents            : Fails without bzr revision id
